Sources/PromotedCore/AnomalyHandling/AnomalyHandler.swift

```diff
@@ -0,0 +1,184 @@
+ */
+class AnomalyHandler: OperationMonitorListener {
+
+  typealias Deps = (
+    ClientConfigSource &
+    OperationMonitorSource &
+    OSLogSource &
+    UIStateSource
+  )
+
+  weak var delegate: AnomalyHandlerDelegate?
+
+  private let config: ClientConfig
+  private unowned let osLog: OSLog?
+  private unowned let uiState: UIState
+
+  private(set) var anomalyCount: Int
+  private var shouldShowModal: Bool
+
+  init(deps: Deps) {
+    assert(deps.clientConfig.loggingAnomalyHandling > .none)
+    config = deps.clientConfig
+    osLog = deps.osLog(category: "AnomalyHandler")
+    uiState = deps.uiState
+    delegate = nil
+    anomalyCount = 0
+    shouldShowModal = true
+    deps.operationMonitor.addOperationMonitorListener(self)
+  }
+
+  func execution(context: Context, didError error: Error) {
+    switch error {
+    case MetricsLoggerError.missingUserIDsInUserMessage:
+      // If an attempt to log a User event without userID AND
+      // logUserID occurs, it goes through this path. Otherwise,
+      // if it's only missing logUserID, it goes through
+      // execution(context:willLogMessage:).
+      triggerAnomalyHandlerResponse(
+        type: .missingLogUserIDInUserMessage,
+        message: nil
+      )
+    default:
+      break
+    }
+  }
+
+  func execution(context: Context, willLogMessage message: Message) {
+    switch message {
+    case let logRequest as Event_LogRequest:
+      analyze(logRequest: logRequest)
+    case let impression as Event_Impression:
+      analyze(impression: impression)
+    case let action as Event_Action:
+      analyze(action: action)
+    case let user as Event_User:
+      // If an attempt to log a User event without a logUserID
+      // occurs (but it has a userID), it goes through this path.
+      // Otherwise, it goes through execution(context:didError:).
+      analyze(user: user)
+    default:
+      break
+    }
+  }
+
+  func analyze(logRequest: Event_LogRequest) {
+    if logRequest.userInfo.logUserID.isEmptyOrWhitespace {
+      triggerAnomalyHandlerResponse(
+        type: .missingLogUserIDInLogRequest,
+        message: logRequest
+      )
+    }
+  }
+
+  func analyze(impression: Event_Impression) {
+    if (
+      impression.sourceType == .delivery &&
+      impression.insertionID.isEmptyOrWhitespace &&
+      impression.contentID.isEmptyOrWhitespace
+    ) {
+      triggerAnomalyHandlerResponse(
+        type: .missingJoinableFieldsInImpression,
+        message: impression
+      )
+    }
+  }
+
+  func analyze(action: Event_Action) {
+    if (
+      action.actionType != .checkout &&
+      action.actionType != .purchase &&
+      action.impressionID.isEmptyOrWhitespace &&
+      action.insertionID.isEmptyOrWhitespace &&
+      action.contentID.isEmptyOrWhitespace
+    ) {
+      triggerAnomalyHandlerResponse(
+        type: .missingJoinableFieldsInAction,
+        message: action
+      )
+    }
+  }
+
+  func analyze(user: Event_User) {
```

Comment:
Side-note - We currently have 2 ways to log Users in our system currently: this way and the User CMS.  This version doesn't provide much value.  It might have a small amount of benefit if the DeliveryLog does not have a userId on it.  The CMS User has a benefit on Delivery.

Not urgent for this PR.  JFYI.

Reply:
Interesting. This method also seems redundant given the `UserInfo` on `LogRequest`.